clients/src/main/java/org/apache/kafka/clients/admin/internals/DescribeShareGroupsHandler.java

```diff
@@ -0,0 +1,381 @@
+import org.apache.kafka.common.requests.DescribeGroupsResponse;
+import org.apache.kafka.common.requests.FindCoordinatorRequest;
+import org.apache.kafka.common.requests.FindCoordinatorRequest.CoordinatorType;
+import org.apache.kafka.common.requests.MetadataResponse;
+import org.apache.kafka.common.requests.ShareGroupDescribeRequest;
+import org.apache.kafka.common.requests.ShareGroupDescribeResponse;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Utils;
+import org.slf4j.Logger;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class DescribeShareGroupsHandler implements AdminApiHandler<CoordinatorKey, ShareGroupDescription> {
+
+  private final boolean includeAuthorizedOperations;
+  private final Logger log;
+  private final AdminApiLookupStrategy<CoordinatorKey> lookupStrategy;
+  private final Set<String> useClassicGroupApi;
+
+  public DescribeShareGroupsHandler(
+      boolean includeAuthorizedOperations,
+      LogContext logContext
+  ) {
+    this.includeAuthorizedOperations = includeAuthorizedOperations;
+    this.log = logContext.logger(DescribeShareGroupsHandler.class);
+    this.lookupStrategy = new CoordinatorStrategy(CoordinatorType.GROUP, logContext);
+    this.useClassicGroupApi = new HashSet<>();
+  }
+
+  private static Set<CoordinatorKey> buildKeySet(Collection<String> groupIds) {
+    return groupIds.stream()
+        .map(CoordinatorKey::byGroupId)
+        .collect(Collectors.toSet());
+  }
+
+  public static AdminApiFuture.SimpleAdminApiFuture<CoordinatorKey, ShareGroupDescription> newFuture(
+      Collection<String> groupIds
+  ) {
+    return AdminApiFuture.forKeys(buildKeySet(groupIds));
+  }
+
+  @Override
+  public String apiName() {
+    return "describeShareGroups";
+  }
+
+  @Override
+  public AdminApiLookupStrategy<CoordinatorKey> lookupStrategy() {
+    return lookupStrategy;
+  }
+
+  @Override
+  public Collection<RequestAndKeys<CoordinatorKey>> buildRequest(int coordinatorId, Set<CoordinatorKey> keys) {
+    Set<CoordinatorKey> newShareGroupKeys = new HashSet<>();
+    Set<CoordinatorKey> oldShareGroupKeys = new HashSet<>();
+    List<String> newShareGroupIds = new ArrayList<>();
+    List<String> oldShareGroupIds = new ArrayList<>();
+
+    keys.forEach(key -> {
+      if (key.type != FindCoordinatorRequest.CoordinatorType.GROUP) {
+        throw new IllegalArgumentException("Invalid group coordinator key " + key +
+            " when building `DescribeGroups` request");
+      }
+
+      // By default, we always try using the new share group describe API.
+      // If it fails, we fail back to using the classic group API.
+      if (useClassicGroupApi.contains(key.idValue)) {
```

Comment:
I'm not convinced by the idea of a "classic group API" for this KIP. It's going to land after everyone is using the new group coordinator.